### ISSUE TITLE
Allow matching on types

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty#4939]
The work in #4927 and preceding PR's showed that it's viable to compute with types, but that the syntax is sometimes a bit clunky. In particular, the following pattern is ubiquitous:

```scala
import scala.typelevel.erasedValue

rewrite erasedValue[T] match {
  case _: A => ...
  ...
  case _: Z => ...
}
```

It would be nice if we had a more streamlined syntax for this. I propose:

```scala
rewrite type T match {
  case A => ...
  ...
  case Z => ...
}
```
Here `T` is an arbitrary type and `A`,...,`Z` are arbitrary type patterns. The construct is semantically equivalent to the first version with `erasedValue`, and can be explained that way. That is, the first case clause where `T` can be made a subtype of the type pattern is chosen. However, the compiler can probably process the second version more efficiently (in terms of compile speed) than the first.

Syntax:

```
Expr1            ::= ...
                     [‘rewrite’] PostfixExpr ‘match’ ‘{’ CaseClauses ‘}’  
                     'type' InfixType ‘match’ ‘{’ TypeCaseClauses ‘}’  
TypeCaseClauses  ::= TypeCaseClause { TypeCaseClause }
TypeCaseClause   ::= 'case' Type [Guard] ‘=>’ Block
```

[EDITED: I had a add `rewrite` in front of `type` since otherwise parsing gets too ambiguous.]

**Comment by jducoeur** [comment]
Just from an end-user POV, I think this is beautifully clear -- I'd love to see it happen...

**Comment by liufengyun** [comment]
We already have match types implementated, I think this is no longer relevant.

http://dotty.epfl.ch/docs/reference/new-types/match-types.html

**Comment by LPTK** [comment]
@liufengyun Aren't match types for defining _types_ that match on types, whereas this proposal is to allow defining _terms_ that match on types?

**Comment by liufengyun** [comment]
@LPTK  Good point. There are many changes in meta-programming since then, we need to reconsider this.
